cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -346,14 +374,396 @@ void AddSortingKernels(VectorKernel base, VectorFunction* func) {
+        return SortIndices(*args[0].make_array(), sort_options, ctx);
+        break;
+      case Datum::CHUNKED_ARRAY:
+        return SortIndices(*args[0].chunked_array(), sort_options, ctx);
+        break;
+      case Datum::RECORD_BATCH: {
+        ARROW_ASSIGN_OR_RAISE(auto table,
+                              Table::FromRecordBatches({args[0].record_batch()}));
+        return SortIndices(*table, sort_options, ctx);
+      } break;
+      case Datum::TABLE:
+        return SortIndices(*args[0].table(), sort_options, ctx);
+        break;
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for sort_indices operation: "
+        "values=",
+        args[0].ToString());
+  }
+
+ private:
+  Result<std::shared_ptr<Array>> SortIndices(const Array& values,
+                                             const SortOptions& options,
+                                             ExecContext* ctx) const {
+    SortOrder order = SortOrder::ASCENDING;
+    if (!options.sort_keys.empty()) {
+      order = options.sort_keys[0].order;
+    }
+    ArraySortOptions array_options(order);
+    ARROW_ASSIGN_OR_RAISE(
+        Datum result, CallFunction("array_sort_indices", {values}, &array_options, ctx));
+    return result.make_array();
+  }
+
+  Result<std::shared_ptr<Array>> SortIndices(const ChunkedArray& values,
+                                             const SortOptions& options,
+                                             ExecContext* ctx) const {
+    SortOrder order = SortOrder::ASCENDING;
+    if (!options.sort_keys.empty()) {
+      order = options.sort_keys[0].order;
+    }
+    ArraySortOptions array_options(order);
+
+    std::shared_ptr<Array> array_values;
+    if (values.num_chunks() == 1) {
+      array_values = values.chunk(0);
+    } else {
+      ARROW_ASSIGN_OR_RAISE(array_values,
+                            Concatenate(values.chunks(), ctx->memory_pool()));
+    }
+    ARROW_ASSIGN_OR_RAISE(Datum result, CallFunction("array_sort_indices", {array_values},
+                                                     &array_options, ctx));
+    return result.make_array();
+  }
+
+  Result<Datum> SortIndices(const Table& table, const SortOptions& options,
+                            ExecContext* ctx) const {
+    auto n_sort_keys = options.sort_keys.size();
+    if (n_sort_keys == 0) {
+      return Status::Invalid("Must specify one or more sort keys");
+    }
+    if (n_sort_keys == 1) {
+      auto chunked_array = table.GetColumnByName(options.sort_keys[0].name);
+      if (!chunked_array) {
+        return Status::Invalid("Nonexistent sort key column: ",
+                               options.sort_keys[0].name);
+      }
+      return SortIndices(*chunked_array, options, ctx);
+    }
+
+    auto out_type = uint64();
+    auto length = table.num_rows();
+    auto buffer_size = BitUtil::BytesForBits(
+        length * std::static_pointer_cast<UInt64Type>(out_type)->bit_width());
```

Comment:
`sizeof(uint64_t)` perhaps?

Reply:
If we want to use `sizeof()`, we need to use `sizeof() * 8`. Because the unit is bit not byte.
Should we use `sizeof() * 8` here?